New 'mysql_servers_ssl_params' settings not working (TLS/SSL error: invalid directory)

**ProxySQL version:**  2.6.0
**OS version:** n/a, Docker image "proxysql/proxysql:2.6.0"|

When using "mysql_servers_ssl_params" instead of the global variables, ProxySQL can't use the given files.

If only given "ssl_ca", the "file open" error is just ignored, resulting in certificate validation issues if using a custom CA. If you use "ssl_cert" and "ssl_key" for cert-based authentication, it fails with "TLS/SSL error: invalid directory".

I've created a docker-compose based demonstration environment; you can find it attached to this report. It contains the needed configuration (proxysql.cnf and certificate/key-files) to reproduce the issue.

[proxysql_ssl_params_test.zip](https://github.com/sysown/proxysql/files/14541581/proxysql_ssl_params_test.zip)
[proxysql_ssl_params_test.tar.bz2.zip](https://github.com/sysown/proxysql/files/14541628/proxysql_ssl_params_test.tar.bz2.zip) (zipped .tar.bz2 to re-create the correct permissions on Linux)

The complete log of a failure from the aforementioned environment:
```
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Using config file /etc/proxysql.cnf
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Current RLIMIT_NOFILE: 1048576
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Using OpenSSL version: OpenSSL 3.2.1 30 Jan 2024
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] No SSL keys/certificates found in datadir (/var/lib/proxysql). Generating new keys/certificates.
db-1                   | 2024-03-08 17:56:21+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.11.7+maria~ubu2204 started.
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Using config file /etc/proxysql.cnf
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Current RLIMIT_NOFILE: 1048576
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Using OpenSSL version: OpenSSL 3.2.1 30 Jan 2024
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] No SSL keys/certificates found in datadir (/var/lib/proxysql). Generating new keys/certificates.
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] ProxySQL version 2.6.0-590-g9878ed3
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Detected OS: Linux c8aeeb5bdb9d 5.15.0-94-generic #104~20.04.1-Ubuntu SMP Tue Jan 16 13:34:09 UTC 2024 x86_64
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] ProxySQL SHA1 checksum: e7e19b20285e13cfbceb88a6a2331afdfcbad8da
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] SSL keys/certificates found in datadir (/var/lib/proxysql): loading them.
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Loaded built-in SQLite3
proxysql-global-1      | Standard ProxySQL MySQL Logger rev. 2.5.0421 -- MySQL_Logger.cpp -- Fri Mar  1 15:22:35 2024
proxysql-global-1      | Standard ProxySQL Cluster rev. 0.4.0906 -- ProxySQL_Cluster.cpp -- Fri Mar  1 15:22:35 2024
proxysql-global-1      | Standard ProxySQL Statistics rev. 1.4.1027 -- ProxySQL_Statistics.cpp -- Fri Mar  1 15:22:35 2024
proxysql-global-1      | Standard ProxySQL HTTP Server Handler rev. 1.4.1031 -- ProxySQL_HTTP_Server.cpp -- Fri Mar  1 15:22:35 2024
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Using UUID: be05a49c-5751-4f0f-85f0-fc030b148ec4 , randomly generated. Writing it to database
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Computed checksum for 'LOAD ADMIN VARIABLES TO RUNTIME' was '0xECFC7190D0FB69B9', with epoch '1709920581'
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Computed checksum for 'LOAD MYSQL VARIABLES TO RUNTIME' was '0x40AE3CA1873ABB26', with epoch '1709920581'
proxysql-global-1      | Standard ProxySQL Admin rev. 2.0.6.0805 -- ProxySQL_Admin.cpp -- Fri Mar  1 15:22:35 2024
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] ProxySQL SHA1 checksum: e7e19b20285e13cfbceb88a6a2331afdfcbad8da
proxysql-global-1      | Standard MySQL Threads Handler rev. 0.2.0902 -- MySQL_Thread.cpp -- Fri Mar  1 15:22:35 2024
proxysql-global-1      | Standard MySQL Authentication rev. 0.2.0902 -- MySQL_Authentication.cpp -- Fri Mar  1 15:22:35 2024
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Computed checksum for 'LOAD MYSQL USERS TO RUNTIME' was '0xD6F3FDE2B2DC1787', with epoch '1709920581'
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Generating runtime mysql servers and mysql servers v2 records.
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Dumping mysql_servers_incoming
proxysql-global-1      | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-global-1      | | hostgroup_id | hostname | port | gtid_port | weight | status | compression | max_connections | max_replication_lag | use_ssl | max_latency_ms | comment |
proxysql-global-1      | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-global-1      | | 10           | db       | 3306 | 0         | 10     | 0      | 0           | 2000            | 0                   | 1       | 0              |         |
proxysql-global-1      | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Dumping mysql_servers LEFT JOIN mysql_servers_incoming
proxysql-global-1      | +-------------+--------------+----------+------+
proxysql-global-1      | | mem_pointer | hostgroup_id | hostname | port |
proxysql-global-1      | +-------------+--------------+----------+------+
proxysql-global-1      | +-------------+--------------+----------+------+
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Dumping mysql_servers JOIN mysql_servers_incoming
proxysql-global-1      | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+-------------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-global-1      | | hostgroup_id | hostname | port | gtid_port | weight | status | compression | max_connections | max_replication_lag | use_ssl | max_latency_ms | comment | mem_pointer | gtid_port | weight | status | compression | max_connections | max_replication_lag | use_ssl | max_latency_ms | comment |
proxysql-global-1      | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+-------------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-global-1      | | 10           | db       | 3306 | 0         | 10     | 0      | 0           | 2000            | 0                   | 1       | 0              |         | 0           | 0         | 10     | 0      | 0           | 2000            | 0                   | 1       | 0              |         |
proxysql-global-1      | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+-------------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Creating new server in HG 10 : db:3306 , gtid_port=0, weight=10, status=0
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] New mysql_group_replication_hostgroups table
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] New mysql_galera_hostgroups table
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] New mysql_aws_aurora_hostgroups table
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] New mysql_hostgroup_attributes table
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] New mysql_servers_ssl_params table
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Checksum for table mysql_servers_v2 is 0x8DA754A126B9F18B
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] New computed global checksum for 'mysql_servers_v2' is '0x528311152709B15E'
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Checksum for table mysql_servers is 0x8DA754A126B9F18B
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Rebuilding 'Hostgroup_Manager_Mapping' due to checksums change - mysql_servers { old: 0x0, new: 0x26B9F18B8DA754A1 }, mysql_replication_hostgroups { old:0x0, new:0x0 }
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] MySQL_HostGroups_Manager::commit() locked for 2ms
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Computed checksum for 'LOAD PROXYSQL SERVERS TO RUNTIME' was '0x0000000000000000', with epoch '1709920581'
proxysql-global-1      | Standard Query Processor rev. 2.0.6.0805 -- Query_Processor.cpp -- Fri Mar  1 15:22:35 2024
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] Computed checksum for 'LOAD MYSQL QUERY RULES TO RUNTIME' was '0xABC239DAF6C84BC1', with epoch '1709920581'
proxysql-global-1      | In memory Standard Query Cache (SQC) rev. 1.2.0905 -- Query_Cache.cpp -- Fri Mar  1 15:22:35 2024
proxysql-global-1      | Standard MySQL Monitor (StdMyMon) rev. 2.0.1226 -- MySQL_Monitor.cpp -- Fri Mar  1 15:22:35 2024
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] For information about products and services visit: https://proxysql.com/
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] For online documentation visit: https://proxysql.com/documentation/
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] For support visit: https://proxysql.com/services/support/
proxysql-global-1      | 2024-03-08 17:56:21 [INFO] For consultancy visit: https://proxysql.com/services/consulting/
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] ProxySQL version 2.6.0-590-g9878ed3
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Detected OS: Linux aa16b632fb71 5.15.0-94-generic #104~20.04.1-Ubuntu SMP Tue Jan 16 13:34:09 UTC 2024 x86_64
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] ProxySQL SHA1 checksum: e7e19b20285e13cfbceb88a6a2331afdfcbad8da
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] SSL keys/certificates found in datadir (/var/lib/proxysql): loading them.
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Loaded built-in SQLite3
proxysql-ssl-params-1  | Standard ProxySQL MySQL Logger rev. 2.5.0421 -- MySQL_Logger.cpp -- Fri Mar  1 15:22:35 2024
proxysql-ssl-params-1  | Standard ProxySQL Cluster rev. 0.4.0906 -- ProxySQL_Cluster.cpp -- Fri Mar  1 15:22:35 2024
proxysql-ssl-params-1  | Standard ProxySQL Statistics rev. 1.4.1027 -- ProxySQL_Statistics.cpp -- Fri Mar  1 15:22:35 2024
proxysql-ssl-params-1  | Standard ProxySQL HTTP Server Handler rev. 1.4.1031 -- ProxySQL_HTTP_Server.cpp -- Fri Mar  1 15:22:35 2024
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Using UUID: 006becf6-9254-4ee6-8885-1bd9813eec55 , randomly generated. Writing it to database
db-1                   | 2024-03-08 17:56:21+00:00 [Warn] [Entrypoint]: /sys/fs/cgroup/rdma:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 12:memory:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 11:perf_event:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 10:freezer:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 9:devices:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 8:net_cls,net_prio:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 7:misc:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 6:pids:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 5:cpu,cpuacct:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 4:cpuset:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 3:hugetlb:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 2:blkio:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 1:name=systemd:/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2
db-1                   | 0::/docker/5f667d9e254b66e4f5f1d7e822d67a845792dd67a9dd97e9d8d8a56cbd320ca2/memory.pressure not writable, functionality unavailable to MariaDB
db-1                   | 2024-03-08 17:56:21+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Computed checksum for 'LOAD ADMIN VARIABLES TO RUNTIME' was '0xECFC7190D0FB69B9', with epoch '1709920581'
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Computed checksum for 'LOAD MYSQL VARIABLES TO RUNTIME' was '0x8B82E9183FA1E428', with epoch '1709920581'
proxysql-ssl-params-1  | Standard ProxySQL Admin rev. 2.0.6.0805 -- ProxySQL_Admin.cpp -- Fri Mar  1 15:22:35 2024
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] ProxySQL SHA1 checksum: e7e19b20285e13cfbceb88a6a2331afdfcbad8da
proxysql-ssl-params-1  | Standard MySQL Threads Handler rev. 0.2.0902 -- MySQL_Thread.cpp -- Fri Mar  1 15:22:35 2024
proxysql-ssl-params-1  | Standard MySQL Authentication rev. 0.2.0902 -- MySQL_Authentication.cpp -- Fri Mar  1 15:22:35 2024
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Computed checksum for 'LOAD MYSQL USERS TO RUNTIME' was '0xD6F3FDE2B2DC1787', with epoch '1709920581'
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Generating runtime mysql servers and mysql servers v2 records.
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Dumping mysql_servers_incoming
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | | hostgroup_id | hostname | port | gtid_port | weight | status | compression | max_connections | max_replication_lag | use_ssl | max_latency_ms | comment |
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | | 10           | db       | 3306 | 0         | 10     | 0      | 0           | 2000            | 0                   | 1       | 0              |         |
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Dumping mysql_servers LEFT JOIN mysql_servers_incoming
proxysql-ssl-params-1  | +-------------+--------------+----------+------+
proxysql-ssl-params-1  | | mem_pointer | hostgroup_id | hostname | port |
proxysql-ssl-params-1  | +-------------+--------------+----------+------+
proxysql-ssl-params-1  | +-------------+--------------+----------+------+
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Dumping mysql_servers JOIN mysql_servers_incoming
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+-------------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | | hostgroup_id | hostname | port | gtid_port | weight | status | compression | max_connections | max_replication_lag | use_ssl | max_latency_ms | comment | mem_pointer | gtid_port | weight | status | compression | max_connections | max_replication_lag | use_ssl | max_latency_ms | comment |
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+-------------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | | 10           | db       | 3306 | 0         | 10     | 0      | 0           | 2000            | 0                   | 1       | 0              |         | 0           | 0         | 10     | 0      | 0           | 2000            | 0                   | 1       | 0              |         |
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+-------------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Creating new server in HG 10 : db:3306 , gtid_port=0, weight=10, status=0
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] New mysql_group_replication_hostgroups table
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] New mysql_galera_hostgroups table
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] New mysql_aws_aurora_hostgroups table
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] New mysql_hostgroup_attributes table
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] New mysql_servers_ssl_params table
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Checksum for table mysql_servers_v2 is 0x8DA754A126B9F18B
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] New computed global checksum for 'mysql_servers_v2' is '0x528311152709B15E'
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Checksum for table mysql_servers is 0x8DA754A126B9F18B
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Rebuilding 'Hostgroup_Manager_Mapping' due to checksums change - mysql_servers { old: 0x0, new: 0x26B9F18B8DA754A1 }, mysql_replication_hostgroups { old:0x0, new:0x0 }
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] MySQL_HostGroups_Manager::commit() locked for 1ms
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Computed checksum for 'LOAD PROXYSQL SERVERS TO RUNTIME' was '0x0000000000000000', with epoch '1709920581'
proxysql-ssl-params-1  | Standard Query Processor rev. 2.0.6.0805 -- Query_Processor.cpp -- Fri Mar  1 15:22:35 2024
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] Computed checksum for 'LOAD MYSQL QUERY RULES TO RUNTIME' was '0xABC239DAF6C84BC1', with epoch '1709920581'
proxysql-ssl-params-1  | In memory Standard Query Cache (SQC) rev. 1.2.0905 -- Query_Cache.cpp -- Fri Mar  1 15:22:35 2024
db-1                   | 2024-03-08 17:56:21+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:10.11.7+maria~ubu2204 started.
proxysql-ssl-params-1  | Standard MySQL Monitor (StdMyMon) rev. 2.0.1226 -- MySQL_Monitor.cpp -- Fri Mar  1 15:22:35 2024
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] For information about products and services visit: https://proxysql.com/
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] For online documentation visit: https://proxysql.com/documentation/
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] For support visit: https://proxysql.com/services/support/
proxysql-ssl-params-1  | 2024-03-08 17:56:21 [INFO] For consultancy visit: https://proxysql.com/services/consulting/
db-1                   | 2024-03-08 17:56:22+00:00 [Note] [Entrypoint]: Initializing database files
proxysql-global-1      | 2024-03-08 17:56:22 main.cpp:146:main_check_latest_version(): [ERROR] curl_easy_perform() failed: SSL connect error
proxysql-ssl-params-1  | 2024-03-08 17:56:22 main.cpp:146:main_check_latest_version(): [ERROR] curl_easy_perform() failed: SSL connect error
db-1                   | 
db-1                   | 
db-1                   | PLEASE REMEMBER TO SET A PASSWORD FOR THE MariaDB root USER !
db-1                   | To do so, start the server, then issue the following command:
db-1                   | 
db-1                   | '/usr/bin/mariadb-secure-installation'
db-1                   | 
db-1                   | which will also give you the option of removing the test
db-1                   | databases and anonymous user created by default.  This is
db-1                   | strongly recommended for production servers.
db-1                   | 
db-1                   | See the MariaDB Knowledgebase at https://mariadb.com/kb
db-1                   | 
db-1                   | Please report any problems at https://mariadb.org/jira
db-1                   | 
db-1                   | The latest information about MariaDB is available at https://mariadb.org/.
db-1                   | 
db-1                   | Consider joining MariaDB's strong and vibrant community:
db-1                   | https://mariadb.org/get-involved/
db-1                   | 
db-1                   | 2024-03-08 17:56:22+00:00 [Note] [Entrypoint]: Database files initialized
db-1                   | 2024-03-08 17:56:22+00:00 [Note] [Entrypoint]: Starting temporary server
db-1                   | 2024-03-08 17:56:22+00:00 [Note] [Entrypoint]: Waiting for server startup
db-1                   | 2024-03-08 17:56:22 0 [Note] Starting MariaDB 10.11.7-MariaDB-1:10.11.7+maria~ubu2204 source revision 87e13722a95af5d9378d990caf48cb6874439347 as process 100
db-1                   | 2024-03-08 17:56:22 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
db-1                   | 2024-03-08 17:56:22 0 [Note] InnoDB: Number of transaction pools: 1
db-1                   | 2024-03-08 17:56:22 0 [Note] InnoDB: Using crc32 + pclmulqdq instructions
db-1                   | 2024-03-08 17:56:22 0 [Note] mariadbd: O_TMPFILE is not supported on /tmp (disabling future attempts)
db-1                   | 2024-03-08 17:56:22 0 [Note] InnoDB: Initializing buffer pool, total size = 128.000MiB, chunk size = 2.000MiB
db-1                   | 2024-03-08 17:56:22 0 [Note] InnoDB: Completed initialization of buffer pool
db-1                   | 2024-03-08 17:56:22 0 [Note] InnoDB: File system buffers for log disabled (block size=512 bytes)
db-1                   | 2024-03-08 17:56:22 0 [Note] InnoDB: End of log at LSN=45518
db-1                   | 2024-03-08 17:56:22 0 [Note] InnoDB: 128 rollback segments are active.
db-1                   | 2024-03-08 17:56:22 0 [Note] InnoDB: Setting file './ibtmp1' size to 12.000MiB. Physically writing the file full; Please wait ...
db-1                   | 2024-03-08 17:56:22 0 [Note] InnoDB: File './ibtmp1' size is now 12.000MiB.
db-1                   | 2024-03-08 17:56:22 0 [Note] InnoDB: log sequence number 45518; transaction id 14
db-1                   | 2024-03-08 17:56:22 0 [Note] Plugin 'FEEDBACK' is disabled.
db-1                   | 2024-03-08 17:56:22 0 [Warning] 'user' entry 'root@5f667d9e254b' ignored in --skip-name-resolve mode.
db-1                   | 2024-03-08 17:56:22 0 [Warning] 'proxies_priv' entry '@% root@5f667d9e254b' ignored in --skip-name-resolve mode.
db-1                   | 2024-03-08 17:56:22 0 [Note] mariadbd: ready for connections.
db-1                   | Version: '10.11.7-MariaDB-1:10.11.7+maria~ubu2204'  socket: '/run/mysqld/mysqld.sock'  port: 0  mariadb.org binary distribution
db-1                   | 2024-03-08 17:56:23+00:00 [Note] [Entrypoint]: Temporary server started.
db-1                   | 2024-03-08 17:56:25+00:00 [Note] [Entrypoint]: Securing system users (equivalent to running mysql_secure_installation)
db-1                   | 
db-1                   | 2024-03-08 17:56:25+00:00 [Note] [Entrypoint]: /usr/local/bin/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/user_setup.sql
db-1                   | 
db-1                   | 
db-1                   | 2024-03-08 17:56:25+00:00 [Note] [Entrypoint]: Stopping temporary server
db-1                   | 2024-03-08 17:56:25 0 [Note] mariadbd (initiated by: unknown): Normal shutdown
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: FTS optimize thread exiting.
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Starting shutdown...
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Dumping buffer pool(s) to /var/lib/mysql/ib_buffer_pool
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Buffer pool(s) dump completed at 240308 17:56:25
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Removed temporary tablespace data file: "./ibtmp1"
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Shutdown completed; log sequence number 47108; transaction id 15
db-1                   | 2024-03-08 17:56:25 0 [Note] mariadbd: Shutdown complete
db-1                   | 
db-1                   | 2024-03-08 17:56:25+00:00 [Note] [Entrypoint]: Temporary server stopped
db-1                   | 
db-1                   | 2024-03-08 17:56:25+00:00 [Note] [Entrypoint]: MariaDB init process done. Ready for start up.
db-1                   | 
db-1                   | 2024-03-08 17:56:25 0 [Note] Starting MariaDB 10.11.7-MariaDB-1:10.11.7+maria~ubu2204 source revision 87e13722a95af5d9378d990caf48cb6874439347 as process 1
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Number of transaction pools: 1
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Using crc32 + pclmulqdq instructions
db-1                   | 2024-03-08 17:56:25 0 [Note] mariadbd: O_TMPFILE is not supported on /tmp (disabling future attempts)
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Initializing buffer pool, total size = 128.000MiB, chunk size = 2.000MiB
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Completed initialization of buffer pool
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: File system buffers for log disabled (block size=512 bytes)
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: End of log at LSN=47108
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: 128 rollback segments are active.
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Setting file './ibtmp1' size to 12.000MiB. Physically writing the file full; Please wait ...
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: File './ibtmp1' size is now 12.000MiB.
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: log sequence number 47108; transaction id 16
db-1                   | 2024-03-08 17:56:25 0 [Note] Plugin 'FEEDBACK' is disabled.
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
db-1                   | 2024-03-08 17:56:25 0 [Warning] You need to use --log-bin to make --expire-logs-days or --binlog-expire-logs-seconds work.
db-1                   | 2024-03-08 17:56:25 0 [Note] InnoDB: Buffer pool(s) load completed at 240308 17:56:25
db-1                   | 2024-03-08 17:56:25 0 [Note] Server socket created on IP: '0.0.0.0'.
db-1                   | 2024-03-08 17:56:25 0 [Note] Server socket created on IP: '::'.
db-1                   | 2024-03-08 17:56:25 0 [Note] mariadbd: ready for connections.
db-1                   | Version: '10.11.7-MariaDB-1:10.11.7+maria~ubu2204'  socket: '/run/mysqld/mysqld.sock'  port: 3306  mariadb.org binary distribution
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Received LOAD MYSQL SERVERS TO RUNTIME command
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Generating runtime mysql servers and mysql servers v2 records.
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Dumping current MySQL Servers structures for hostgroup ALL
proxysql-ssl-params-1  | HID: 10 , address: db , port: 3306 , gtid_port: 0 , weight: 10 , status: ONLINE , max_connections: 2000 , max_replication_lag: 0 , use_ssl: 1 , max_latency_ms: 0 , comment: 
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Dumping mysql_servers: ALL
proxysql-ssl-params-1  | +-----+----------+------+------+--------+--------+-----+-----------+---------+-----+---------+---------+-----------------+
proxysql-ssl-params-1  | | hid | hostname | port | gtid | weight | status | cmp | max_conns | max_lag | ssl | max_lat | comment | mem_pointer     |
proxysql-ssl-params-1  | +-----+----------+------+------+--------+--------+-----+-----------+---------+-----+---------+---------+-----------------+
proxysql-ssl-params-1  | | 10  | db       | 3306 | 0    | 10     | 0      | 0   | 2000      | 0       | 1   | 0       |         | 140242012450368 |
proxysql-ssl-params-1  | +-----+----------+------+------+--------+--------+-----+-----------+---------+-----+---------+---------+-----------------+
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Dumping mysql_servers_incoming
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | | hostgroup_id | hostname | port | gtid_port | weight | status | compression | max_connections | max_replication_lag | use_ssl | max_latency_ms | comment |
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | | 10           | db       | 3306 | 0         | 10     | 0      | 0           | 2000            | 0                   | 1       | 0              |         |
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Dumping mysql_servers LEFT JOIN mysql_servers_incoming
proxysql-ssl-params-1  | +-------------+--------------+----------+------+
proxysql-ssl-params-1  | | mem_pointer | hostgroup_id | hostname | port |
proxysql-ssl-params-1  | +-------------+--------------+----------+------+
proxysql-ssl-params-1  | +-------------+--------------+----------+------+
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Dumping mysql_servers JOIN mysql_servers_incoming
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+-------------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | | hostgroup_id | hostname | port | gtid_port | weight | status | compression | max_connections | max_replication_lag | use_ssl | max_latency_ms | comment | mem_pointer | gtid_port | weight | status | compression | max_connections | max_replication_lag | use_ssl | max_latency_ms | comment |
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+-------------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | +--------------+----------+------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+-------------+-----------+--------+--------+-------------+-----------------+---------------------+---------+----------------+---------+
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] New mysql_replication_hostgroups table
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] New mysql_group_replication_hostgroups table
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] New mysql_galera_hostgroups table
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] New mysql_aws_aurora_hostgroups table
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] New mysql_hostgroup_attributes table
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] New mysql_servers_ssl_params table
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Loading MySQL Server SSL Params for (db,3306,)
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Checksum for table mysql_servers_v2 is 0x8DA754A126B9F18B
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Checksum for table mysql_servers_ssl_params is 0x41039902D46F80AA
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] New computed global checksum for 'mysql_servers_v2' is '0xEB18B5B6E7B68F7D'
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Dumping current MySQL Servers structures for hostgroup ALL
proxysql-ssl-params-1  | HID: 10 , address: db , port: 3306 , gtid_port: 0 , weight: 10 , status: ONLINE , max_connections: 2000 , max_replication_lag: 0 , use_ssl: 1 , max_latency_ms: 0 , comment: 
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Dumping mysql_servers: ALL
proxysql-ssl-params-1  | +-----+----------+------+------+--------+--------+-----+-----------+---------+-----+---------+---------+-----------------+
proxysql-ssl-params-1  | | hid | hostname | port | gtid | weight | status | cmp | max_conns | max_lag | ssl | max_lat | comment | mem_pointer     |
proxysql-ssl-params-1  | +-----+----------+------+------+--------+--------+-----+-----------+---------+-----+---------+---------+-----------------+
proxysql-ssl-params-1  | | 10  | db       | 3306 | 0    | 10     | 0      | 0   | 2000      | 0       | 1   | 0       |         | 140242012450368 |
proxysql-ssl-params-1  | +-----+----------+------+------+--------+--------+-----+-----------+---------+-----+---------+---------+-----------------+
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] Checksum for table mysql_servers is 0x8DA754A126B9F18B
proxysql-ssl-params-1  | 2024-03-08 17:56:30 [INFO] MySQL_HostGroups_Manager::commit() locked for 4ms
proxysql-ssl-params-1  | 2024-03-08 17:56:46 mysql_connection.cpp:1164:handler(): [ERROR] Failed to mysql_real_connect() on 10:db:3306 , FD (Conn:40 , MyDS:40) , 2026: TLS/SSL error: invalid directory. SSL Params: /ssl/ca.crt , /ssl/client.crt , /ssl/client.key ,  ,  ,  ,  , 
db-1                   | 2024-03-08 17:56:46 3 [Warning] Aborted connection 3 to db: 'unconnected' user: 'unauthenticated' host: '172.29.0.4' (This connection closed normally without authentication)
db-1                   | 2024-03-08 17:56:46 4 [Warning] Aborted connection 4 to db: 'unconnected' user: 'unauthenticated' host: '172.29.0.4' (This connection closed normally without authentication)
proxysql-ssl-params-1  | 2024-03-08 17:56:46 mysql_connection.cpp:1164:handler(): [ERROR] Failed to mysql_real_connect() on 10:db:3306 , FD (Conn:40 , MyDS:40) , 2026: TLS/SSL error: invalid directory. SSL Params: /ssl/ca.crt , /ssl/client.crt , /ssl/client.key ,  ,  ,  ,  , 
db-1                   | 2024-03-08 17:56:46 5 [Warning] Aborted connection 5 to db: 'unconnected' user: 'unauthenticated' host: '172.29.0.4' (This connection closed normally without authentication)
proxysql-ssl-params-1  | 2024-03-08 17:56:46 mysql_connection.cpp:1164:handler(): [ERROR] Failed to mysql_real_connect() on 10:db:3306 , FD (Conn:40 , MyDS:40) , 2026: TLS/SSL error: invalid directory. SSL Params: /ssl/ca.crt , /ssl/client.crt , /ssl/client.key ,  ,  ,  ,  , 
proxysql-ssl-params-1  | 2024-03-08 17:56:46 MySQL_HostGroups_Manager.cpp:853:connect_error(): [ERROR] Shunning server db:3306 with 5 errors/sec. Shunning for 10 seconds
proxysql-ssl-params-1  | 2024-03-08 17:56:46 MySQL_HostGroups_Manager.cpp:3103:get_random_MySrvC(): [ERROR] Hostgroup 10 has no servers available! Checking servers shunned for more than 1 second
proxysql-ssl-params-1  | 2024-03-08 17:56:48 MySQL_HostGroups_Manager.cpp:3103:get_random_MySrvC(): [ERROR] Hostgroup 10 has no servers available! Checking servers shunned for more than 1 second
proxysql-ssl-params-1  | 2024-03-08 17:56:48 mysql_connection.cpp:1164:handler(): [ERROR] Failed to mysql_real_connect() on 10:db:3306 , FD (Conn:40 , MyDS:40) , 2026: TLS/SSL error: invalid directory. SSL Params: /ssl/ca.crt , /ssl/client.crt , /ssl/client.key ,  ,  ,  ,  , 
db-1                   | 2024-03-08 17:56:48 6 [Warning] Aborted connection 6 to db: 'unconnected' user: 'unauthenticated' host: '172.29.0.4' (This connection closed normally without authentication)
db-1                   | 2024-03-08 17:56:48 7 [Warning] Aborted connection 7 to db: 'unconnected' user: 'unauthenticated' host: '172.29.0.4' (This connection closed normally without authentication)
proxysql-ssl-params-1  | 2024-03-08 17:56:48 mysql_connection.cpp:1164:handler(): [ERROR] Failed to mysql_real_connect() on 10:db:3306 , FD (Conn:40 , MyDS:40) , 2026: TLS/SSL error: invalid directory. SSL Params: /ssl/ca.crt , /ssl/client.crt , /ssl/client.key ,  ,  ,  ,  , 
db-1                   | 2024-03-08 17:56:48 8 [Warning] Aborted connection 8 to db: 'unconnected' user: 'unauthenticated' host: '172.29.0.4' (This connection closed normally without authentication)
proxysql-ssl-params-1  | 2024-03-08 17:56:48 mysql_connection.cpp:1164:handler(): [ERROR] Failed to mysql_real_connect() on 10:db:3306 , FD (Conn:40 , MyDS:40) , 2026: TLS/SSL error: invalid directory. SSL Params: /ssl/ca.crt , /ssl/client.crt , /ssl/client.key ,  ,  ,  ,  , 
proxysql-ssl-params-1  | 2024-03-08 17:56:48 MySQL_HostGroups_Manager.cpp:853:connect_error(): [ERROR] Shunning server db:3306 with 5 errors/sec. Shunning for 10 seconds
proxysql-ssl-params-1  | 2024-03-08 17:56:50 MySQL_HostGroups_Manager.cpp:3103:get_random_MySrvC(): [ERROR] Hostgroup 10 has no servers available! Checking servers shunned for more than 1 second
proxysql-ssl-params-1  | 2024-03-08 17:56:50 mysql_connection.cpp:1164:handler(): [ERROR] Failed to mysql_real_connect() on 10:db:3306 , FD (Conn:40 , MyDS:40) , 2026: TLS/SSL error: invalid directory. SSL Params: /ssl/ca.crt , /ssl/client.crt , /ssl/client.key ,  ,  ,  ,  , 
db-1                   | 2024-03-08 17:56:50 9 [Warning] Aborted connection 9 to db: 'unconnected' user: 'unauthenticated' host: '172.29.0.4' (This connection closed normally without authentication)
proxysql-ssl-params-1  | 2024-03-08 17:56:50 mysql_connection.cpp:1164:handler(): [ERROR] Failed to mysql_real_connect() on 10:db:3306 , FD (Conn:40 , MyDS:40) , 2026: TLS/SSL error: invalid directory. SSL Params: /ssl/ca.crt , /ssl/client.crt , /ssl/client.key ,  ,  ,  ,  , 
db-1                   | 2024-03-08 17:56:50 10 [Warning] Aborted connection 10 to db: 'unconnected' user: 'unauthenticated' host: '172.29.0.4' (This connection closed normally without authentication)
proxysql-ssl-params-1  | 2024-03-08 17:56:50 mysql_connection.cpp:1164:handler(): [ERROR] Failed to mysql_real_connect() on 10:db:3306 , FD (Conn:40 , MyDS:40) , 2026: TLS/SSL error: invalid directory. SSL Params: /ssl/ca.crt , /ssl/client.crt , /ssl/client.key ,  ,  ,  ,  , 
proxysql-ssl-params-1  | 2024-03-08 17:56:50 MySQL_HostGroups_Manager.cpp:853:connect_error(): [ERROR] Shunning server db:3306 with 5 errors/sec. Shunning for 10 seconds
db-1                   | 2024-03-08 17:56:50 11 [Warning] Aborted connection 11 to db: 'unconnected' user: 'unauthenticated' host: '172.29.0.4' (This connection closed normally without authentication)
proxysql-ssl-params-1  | 2024-03-08 17:56:52 MySQL_HostGroups_Manager.cpp:3103:get_random_MySrvC(): [ERROR] Hostgroup 10 has no servers available! Checking servers shunned for more than 1 second
db-1                   | 2024-03-08 17:56:52 12 [Warning] Aborted connection 12 to db: 'unconnected' user: 'unauthenticated' host: '172.29.0.4' (This connection closed normally without authentication)
proxysql-ssl-params-1  | 2024-03-08 17:56:52 mysql_connection.cpp:1164:handler(): [ERROR] Failed to mysql_real_connect() on 10:db:3306 , FD (Conn:40 , MyDS:40) , 2026: TLS/SSL error: invalid directory. SSL Params: /ssl/ca.crt , /ssl/client.crt , /ssl/client.key ,  ,  ,  ,  , 
proxysql-ssl-params-1  | 2024-03-08 17:56:52 mysql_connection.cpp:1164:handler(): [ERROR] Failed to mysql_real_connect() on 10:db:3306 , FD (Conn:40 , MyDS:40) , 2026: TLS/SSL error: invalid directory. SSL Params: /ssl/ca.crt , /ssl/client.crt , /ssl/client.key ,  ,  ,  ,  , 
db-1                   | 2024-03-08 17:56:52 13 [Warning] Aborted connection 13 to db: 'unconnected' user: 'unauthenticated' host: '172.29.0.4' (This connection closed normally without authentication)
```

--- a/include/mysql_connection.h
+++ b/include/mysql_connection.h
@@ -253,5 +253,6 @@ class MySQL_Connection {
 	bool requires_CHANGE_USER(const MySQL_Connection *client_conn);
 	unsigned int number_of_matching_session_variables(const MySQL_Connection *client_conn, unsigned int& not_matching);
 	unsigned long get_mysql_thread_id() { return mysql ? mysql->thread_id : 0; }
+	static void set_ssl_params(MYSQL *mysql, MySQLServers_SslParams *ssl_params);
 };
 #endif /* __CLASS_MYSQL_CONNECTION_H */

--- a/lib/MySQL_Monitor.cpp
+++ b/lib/MySQL_Monitor.cpp
@@ -1525,15 +1525,10 @@ __exit_set_wait_timeout:
 bool MySQL_Monitor_State_Data::create_new_connection() {
 		mysql=mysql_init(NULL);
 		assert(mysql);
-		if (use_ssl) {
-			mysql_ssl_set(mysql,
-					mysql_thread___ssl_p2s_key,
-					mysql_thread___ssl_p2s_cert,
-					mysql_thread___ssl_p2s_ca,
-					mysql_thread___ssl_p2s_capath,
-					mysql_thread___ssl_p2s_cipher);
-			mysql_options(mysql, MYSQL_OPT_SSL_CRL, mysql_thread___ssl_p2s_crl);
-			mysql_options(mysql, MYSQL_OPT_SSL_CRLPATH, mysql_thread___ssl_p2s_crlpath);
+		MySQLServers_SslParams * ssl_params = NULL;
+		if (use_ssl && port) {
+			ssl_params = MyHGM->get_Server_SSL_Params(hostname, port, mysql_thread___monitor_username);
+			MySQL_Connection::set_ssl_params(mysql,ssl_params);
 			mysql_options(mysql, MARIADB_OPT_SSL_KEYLOG_CALLBACK, (void*)proxysql_keylog_write_line_callback);
 		}
 		unsigned int timeout=mysql_thread___monitor_connect_timeout/1000;
@@ -1551,6 +1546,13 @@ bool MySQL_Monitor_State_Data::create_new_connection() {
 			mysql_error_msg=strdup(mysql_error(mysql));
 			int myerrno=mysql_errno(mysql);
 			MyHGM->p_update_mysql_error_counter(p_mysql_error_type::proxysql, hostgroup_id, hostname, port, myerrno);
+			if (ssl_params != NULL && myerrno == 2026) {
+				proxy_error("Failed to connect to server %s:%d . SSL Params: %s , %s , %s , %s , %s , %s , %s , %s\n",
+					( port ? hostname : "localhost" ) , port ,
+					ssl_params->ssl_ca.c_str() , ssl_params->ssl_cert.c_str() , ssl_params->ssl_key.c_str() , ssl_params->ssl_capath.c_str() ,
+					ssl_params->ssl_crl.c_str() , ssl_params->ssl_crlpath.c_str() , ssl_params->ssl_cipher.c_str() , ssl_params->tls_version.c_str()
+				);
+			}
 			if (myerrno < 2000) {
 				mysql_close(mysql);
 			} else {

--- a/lib/MySQL_Session.cpp
+++ b/lib/MySQL_Session.cpp
@@ -222,6 +222,9 @@ void* kill_query_thread(void *arg) {
 	std::unique_ptr<MySQL_Thread> mysql_thr(new MySQL_Thread());
 	mysql_thr->curtime=monotonic_time();
 	mysql_thr->refresh_variables();
+
+	MySQLServers_SslParams * ssl_params = NULL;
+
 	MYSQL *mysql=mysql_init(NULL);
 	if (!mysql) {
 		goto __exit_kill_query_thread;
@@ -230,15 +233,10 @@ void* kill_query_thread(void *arg) {
 	mysql_options4(mysql, MYSQL_OPT_CONNECT_ATTR_ADD, "program_name", "proxysql_killer");
 	mysql_options4(mysql, MYSQL_OPT_CONNECT_ATTR_ADD, "_server_host", ka->hostname);
 
+
 	if (ka->use_ssl && ka->port) {
-		mysql_ssl_set(mysql, 
-			mysql_thread___ssl_p2s_key,
-			mysql_thread___ssl_p2s_cert,
-			mysql_thread___ssl_p2s_ca,
-			mysql_thread___ssl_p2s_capath,
-			mysql_thread___ssl_p2s_cipher);
-		mysql_options(mysql, MYSQL_OPT_SSL_CRL, mysql_thread___ssl_p2s_crl);
-		mysql_options(mysql, MYSQL_OPT_SSL_CRLPATH, mysql_thread___ssl_p2s_crlpath);
+		ssl_params = MyHGM->get_Server_SSL_Params(ka->hostname, ka->port, ka->username);
+		MySQL_Connection::set_ssl_params(mysql,ssl_params);
 		mysql_options(mysql, MARIADB_OPT_SSL_KEYLOG_CALLBACK, (void*)proxysql_keylog_write_line_callback);
 	}
 
@@ -275,7 +273,16 @@ void* kill_query_thread(void *arg) {
 		ret=mysql_real_connect(mysql,"localhost",ka->username,ka->password,NULL,0,ka->hostname,0);
 	}
 	if (!ret) {
-		proxy_error("Failed to connect to server %s:%d to run KILL %s %lu: Error: %s\n" , ka->hostname, ka->port, ( ka->kill_type==KILL_QUERY ? "QUERY" : "CONNECTION" ) , ka->id, mysql_error(mysql));
+		int myerr = mysql_errno(mysql);
+		if (ssl_params != NULL && myerr == 2026) {
+			proxy_error("Failed to connect to server %s:%d to run KILL %s %lu.  SSL Params: %s , %s , %s , %s , %s , %s , %s , %s\n",
+				ka->hostname, ka->port, ( ka->kill_type==KILL_QUERY ? "QUERY" : "CONNECTION" ) , ka->id,
+				ssl_params->ssl_ca.c_str() , ssl_params->ssl_cert.c_str() , ssl_params->ssl_key.c_str() , ssl_params->ssl_capath.c_str() ,
+				ssl_params->ssl_crl.c_str() , ssl_params->ssl_crlpath.c_str() , ssl_params->ssl_cipher.c_str() , ssl_params->tls_version.c_str()
+			);
+		} else {
+			proxy_error("Failed to connect to server %s:%d to run KILL %s %lu: Error: %s\n" , ka->hostname, ka->port, ( ka->kill_type==KILL_QUERY ? "QUERY" : "CONNECTION" ) , ka->id, mysql_error(mysql));
+		}
 		MyHGM->p_update_mysql_error_counter(p_mysql_error_type::mysql, ka->hid, ka->hostname, ka->port, mysql_errno(mysql));
 		goto __exit_kill_query_thread;
 	}
@@ -300,6 +307,10 @@ __exit_kill_query_thread:
 	if (mysql)
 		mysql_close(mysql);
 	delete ka;
+	if (ssl_params != NULL) {
+		delete ssl_params;
+		ssl_params = NULL;
+	}
 	return NULL;
 }
 

--- a/lib/MySQL_Session.cpp
+++ b/lib/MySQL_Session.cpp
@@ -223,6 +223,10 @@ void* kill_query_thread(void *arg) {
 	mysql_thr->curtime=monotonic_time();
 	mysql_thr->refresh_variables();
 	MYSQL *mysql=mysql_init(NULL);
+	if (!mysql) {
+		goto __exit_kill_query_thread;
+	}
+
 	mysql_options4(mysql, MYSQL_OPT_CONNECT_ATTR_ADD, "program_name", "proxysql_killer");
 	mysql_options4(mysql, MYSQL_OPT_CONNECT_ATTR_ADD, "_server_host", ka->hostname);
 
@@ -238,9 +242,6 @@ void* kill_query_thread(void *arg) {
 		mysql_options(mysql, MARIADB_OPT_SSL_KEYLOG_CALLBACK, (void*)proxysql_keylog_write_line_callback);
 	}
 
-	if (!mysql) {
-		goto __exit_kill_query_thread;
-	}
 	MYSQL *ret;
 	if (ka->port) {
 		switch (ka->kill_type) {

--- a/lib/ProxySQL_Config.cpp
+++ b/lib/ProxySQL_Config.cpp
@@ -1087,6 +1087,59 @@ int ProxySQL_Config::Read_MySQL_Servers_from_configfile() {
 			rows++;
 		}
 	}
+	if (root.exists("mysql_servers_ssl_params")==true) { // mysql_servers_ssl_params
+		const Setting &mysql_servers_ssl_params = root["mysql_servers_ssl_params"];
+		int count = mysql_servers_ssl_params.getLength();
+		char *q=(char *)"INSERT OR REPLACE INTO mysql_servers_ssl_params (hostname, port, username, ssl_ca, ssl_cert, ssl_key, ssl_capath, ssl_crl, ssl_crlpath, ssl_cipher, tls_version, comment) VALUES ('%s', %d, '%s', '%s', '%s', '%s', '%s', '%s', '%s', '%s', '%s', '%s')";
+		for (i=0; i< count; i++) {
+			const Setting &line = mysql_servers_ssl_params[i];
+			string hostname = "";
+			int port = 3306;
+			string username = "";
+			string ssl_ca = "";
+			string ssl_cert = "";
+			string ssl_key = "";
+			string ssl_capath = "";
+			string ssl_crl = "";
+			string ssl_crlpath = "";
+			string ssl_cipher = "";
+			string tls_version = "";
+			std::string comment="";
+			if (line.lookupValue("hostname", hostname)==false) {
+				proxy_error("Admin: detected a mysql_servers_ssl_params in config file without a mandatory hostname\n");
+				continue;
+			}
+			line.lookupValue("port", port);
+			line.lookupValue("username", username);
+			line.lookupValue("ssl_ca", ssl_ca);
+			line.lookupValue("ssl_cert", ssl_cert);
+			line.lookupValue("ssl_key", ssl_key);
+			line.lookupValue("ssl_capath", ssl_capath);
+			line.lookupValue("ssl_crl", ssl_crl);
+			line.lookupValue("ssl_crlpath", ssl_crlpath);
+			line.lookupValue("ssl_cipher", ssl_cipher);
+			line.lookupValue("tls_version", tls_version);
+			line.lookupValue("comment", comment);
+			char *o1=strdup(comment.c_str());
+			char *o=escape_string_single_quotes(o1, false);
+			char *query=(char *)malloc(
+				strlen(q)
+				+ hostname.length() + username.length()
+				+ ssl_ca.length() + ssl_cert.length() + ssl_key.length() + ssl_capath.length()
+				+ ssl_crl.length() + ssl_crlpath.length() + ssl_cipher.length() + tls_version.length()
+				+ strlen(o) + 32);
+			sprintf(query, q,
+				hostname.c_str() , port , username.c_str() ,
+				ssl_ca.c_str() , ssl_cert.c_str() , ssl_key.c_str() , ssl_capath.c_str() ,
+				ssl_crl.c_str() , ssl_crlpath.c_str() , ssl_cipher.c_str() , tls_version.c_str() ,
+				o);
+			admindb->execute(query);
+			if (o!=o1) free(o);
+			free(o1);
+			free(query);
+			rows++;
+		}
+	}
         if (root.exists("mysql_group_replication_hostgroups")==true) {
                 const Setting &mysql_group_replication_hostgroups = root["mysql_group_replication_hostgroups"];
                 int count = mysql_group_replication_hostgroups.getLength();

--- a/lib/mysql_connection.cpp
+++ b/lib/mysql_connection.cpp
@@ -750,28 +750,7 @@ void MySQL_Connection::connect_start() {
 			ssl_params = NULL;
 		}
 		ssl_params = MyHGM->get_Server_SSL_Params(parent->address, parent->port, userinfo->username);
-		if (ssl_params == NULL) {
-			mysql_ssl_set(mysql,
-					mysql_thread___ssl_p2s_key,
-					mysql_thread___ssl_p2s_cert,
-					mysql_thread___ssl_p2s_ca,
-					mysql_thread___ssl_p2s_capath,
-					mysql_thread___ssl_p2s_cipher);
-			mysql_options(mysql, MYSQL_OPT_SSL_CRL, mysql_thread___ssl_p2s_crl);
-			mysql_options(mysql, MYSQL_OPT_SSL_CRLPATH, mysql_thread___ssl_p2s_crlpath);
-		} else {
-			mysql_ssl_set(mysql,
-					( ssl_params->ssl_key.length()    > 0 ? ssl_params->ssl_key.c_str()    : NULL ) ,
-					( ssl_params->ssl_cert.length()   > 0 ? ssl_params->ssl_cert.c_str()   : NULL ) ,
-					( ssl_params->ssl_ca.length()     > 0 ? ssl_params->ssl_ca.c_str()     : NULL ) ,
-					( ssl_params->ssl_capath.length() > 0 ? ssl_params->ssl_capath.c_str() : NULL ) ,
-					( ssl_params->ssl_cipher.length() > 0 ? ssl_params->ssl_cipher.c_str() : NULL )
-			);
-			mysql_options(mysql, MYSQL_OPT_SSL_CRL,
-				( ssl_params->ssl_crl.length() > 0 ? ssl_params->ssl_crl.c_str() : NULL ) );
-			mysql_options(mysql, MYSQL_OPT_SSL_CRLPATH,
-				( ssl_params->ssl_crlpath.length() > 0 ? ssl_params->ssl_crlpath.c_str() : NULL ) );
-		}
+		MySQL_Connection::set_ssl_params(mysql, ssl_params);
 		mysql_options(mysql, MARIADB_OPT_SSL_KEYLOG_CALLBACK, (void*)proxysql_keylog_write_line_callback);
 	}
 	unsigned int timeout= 1;
@@ -2977,4 +2956,29 @@ bool MySQL_Connection::get_gtid(char *buff, uint64_t *trx_id) {
 		}
 	}
 	return ret;
+}
+
+void MySQL_Connection::set_ssl_params(MYSQL *mysql, MySQLServers_SslParams *ssl_params) {
+	if (ssl_params == NULL) {
+		mysql_ssl_set(mysql,
+				mysql_thread___ssl_p2s_key,
+				mysql_thread___ssl_p2s_cert,
+				mysql_thread___ssl_p2s_ca,
+				mysql_thread___ssl_p2s_capath,
+				mysql_thread___ssl_p2s_cipher);
+		mysql_options(mysql, MYSQL_OPT_SSL_CRL, mysql_thread___ssl_p2s_crl);
+		mysql_options(mysql, MYSQL_OPT_SSL_CRLPATH, mysql_thread___ssl_p2s_crlpath);
+	} else {
+		mysql_ssl_set(mysql,
+				( ssl_params->ssl_key.length()    > 0 ? ssl_params->ssl_key.c_str()    : NULL ) ,
+				( ssl_params->ssl_cert.length()   > 0 ? ssl_params->ssl_cert.c_str()   : NULL ) ,
+				( ssl_params->ssl_ca.length()     > 0 ? ssl_params->ssl_ca.c_str()     : NULL ) ,
+				( ssl_params->ssl_capath.length() > 0 ? ssl_params->ssl_capath.c_str() : NULL ) ,
+				( ssl_params->ssl_cipher.length() > 0 ? ssl_params->ssl_cipher.c_str() : NULL )
+		);
+		mysql_options(mysql, MYSQL_OPT_SSL_CRL,
+			( ssl_params->ssl_crl.length() > 0 ? ssl_params->ssl_crl.c_str() : NULL ) );
+		mysql_options(mysql, MYSQL_OPT_SSL_CRLPATH,
+			( ssl_params->ssl_crlpath.length() > 0 ? ssl_params->ssl_crlpath.c_str() : NULL ) );
+	}
 }

--- a/lib/mysql_connection.cpp
+++ b/lib/mysql_connection.cpp
@@ -761,14 +761,16 @@ void MySQL_Connection::connect_start() {
 			mysql_options(mysql, MYSQL_OPT_SSL_CRLPATH, mysql_thread___ssl_p2s_crlpath);
 		} else {
 			mysql_ssl_set(mysql,
-					ssl_params->ssl_key.c_str(),
-					ssl_params->ssl_cert.c_str(),
-					ssl_params->ssl_ca.c_str(),
-					ssl_params->ssl_capath.c_str(),
-					ssl_params->ssl_cipher.c_str()
+					( ssl_params->ssl_key.length()    > 0 ? ssl_params->ssl_key.c_str()    : NULL ) ,
+					( ssl_params->ssl_cert.length()   > 0 ? ssl_params->ssl_cert.c_str()   : NULL ) ,
+					( ssl_params->ssl_ca.length()     > 0 ? ssl_params->ssl_ca.c_str()     : NULL ) ,
+					( ssl_params->ssl_capath.length() > 0 ? ssl_params->ssl_capath.c_str() : NULL ) ,
+					( ssl_params->ssl_cipher.length() > 0 ? ssl_params->ssl_cipher.c_str() : NULL )
 			);
-			mysql_options(mysql, MYSQL_OPT_SSL_CRL, ssl_params->ssl_crl.c_str());
-			mysql_options(mysql, MYSQL_OPT_SSL_CRLPATH, ssl_params->ssl_crlpath.c_str());
+			mysql_options(mysql, MYSQL_OPT_SSL_CRL,
+				( ssl_params->ssl_crl.length() > 0 ? ssl_params->ssl_crl.c_str() : NULL ) );
+			mysql_options(mysql, MYSQL_OPT_SSL_CRLPATH,
+				( ssl_params->ssl_crlpath.length() > 0 ? ssl_params->ssl_crlpath.c_str() : NULL ) );
 		}
 		mysql_options(mysql, MARIADB_OPT_SSL_KEYLOG_CALLBACK, (void*)proxysql_keylog_write_line_callback);
 	}

